shell/shared/grain.js

```diff
@@ -511,6 +511,62 @@ if (Meteor.isClient) {
             currentGrainId + event.data.setPath);
       } else if (event.data.setTitle) {
         Session.set("grainFrameTitle", event.data.setTitle);
+      } else if (event.data.renderTemplate) {
+        // Request creation of a single-use template with a privileged API token.
+        // Why?  Apps should not be able to obtain capabilities-as-keys to
+        // themselves directly, because those can be leaked through an
+        // arbitrary bit stream or covert channel.  However, apps often need a
+        // way to provide instructions to users to copy/paste with some
+        // privileged token contained within.  By providing this templating in
+        // the platform, we can ensure that the token is only visible to the
+        // shell's origin.
+        var call = event.data.renderTemplate;
+        check(call, Object);
+        var rpcId = call.rpcId;
+        check(rpcId, String);
+        var template = call.template;
+        check(template, String);
+        var assignment = undefined;
+        if (call.roleId) {
+          check(call.roleId, Match.Integer);
+          assignment = {roleId: call.roleId};
+        } else {
+          assignment = {none: null};
+        }
+        var petname = undefined;
+        if (call.petname) {
+          check(call.petname, String);
+          petname = call.petname;
+        } else {
+          petname = "connected external app";
+        }
+        // Tokens expire by default in 5 minutes from generation date
+        var selfDestructTime = Date.now() + (5 * 60 * 1000);
+        Meteor.call("newApiToken", currentGrainId, petname, assignment, false,
+                    selfDestructTime, function (error, result) {
+          if (error) {
+            //Session.set("api-token-" + currentGrainId, undefined);
+            window.alert("Failed to create token for offer template.\n" + error);
+            console.error(error.stack);
+          } else {
+            var tokenId = result.token;
+            // Generate random key id2.
+            var id2 = Random.secret();
+            // Store apitoken id1 and template in local storage in the offer
+            // template namespace under key id2.
+            var key = "offerTemplate" + id2;
+            var renderedTemplate = template.replace("$API_TOKEN", tokenId)
+                                           .replace("$API_HOST", makeWildcardHost("api"));
+            sessionStorage.setItem(key, JSON.stringify({
+                "renderedTemplate": renderedTemplate,
+                "expiry": selfDestructTime
```

Comment:
It looks like the code in offer-template.html expects an "expires" field, not an "expiry" field.